cli/context/store/store.go

```diff
@@ -282,37 +304,108 @@ func Import(name string, s Writer, reader io.Reader) error {
 			if err != nil {
 				return err
 			}
-			var meta Metadata
-			if err := json.Unmarshal(data, &meta); err != nil {
+			meta, err := parseMetadata(data, name)
+			if err != nil {
 				return err
 			}
-			meta.Name = name
 			if err := s.CreateOrUpdate(meta); err != nil {
 				return err
 			}
 		} else if strings.HasPrefix(hdr.Name, "tls/") {
-			relative := strings.TrimPrefix(hdr.Name, "tls/")
-			parts := strings.SplitN(relative, "/", 2)
-			if len(parts) != 2 {
-				return errors.New("archive format is invalid")
-			}
-			endpointName := parts[0]
-			fileName := parts[1]
 			data, err := ioutil.ReadAll(tr)
 			if err != nil {
 				return err
 			}
-			if _, ok := tlsData.Endpoints[endpointName]; !ok {
-				tlsData.Endpoints[endpointName] = EndpointTLSData{
-					Files: map[string][]byte{},
-				}
+			if err := importEndpointTLS(&tlsData, hdr.Name, data); err != nil {
+				return err
 			}
-			tlsData.Endpoints[endpointName].Files[fileName] = data
 		}
 	}
 	return s.ResetTLSMaterial(name, &tlsData)
 }
 
+func importZip(name string, s Writer, reader io.Reader) error {
+	body, err := limitedReadAll(reader, maxFileSize)
+	if err != nil {
+		return err
+	}
+	zr, err := zip.NewReader(bytes.NewReader(body), int64(len(body)))
+	if err != nil {
+		return err
+	}
+	tlsData := ContextTLSData{
+		Endpoints: map[string]EndpointTLSData{},
+	}
+	for _, zf := range zr.File {
+		fi := zf.FileInfo()
+		if fi.IsDir() {
+			// skip this entry, only taking files into account
+			continue
+		}
+		if zf.Name == metaFile {
```

Comment:
I went for just duplicating this import, instead of trying to abstract it; it felt like the abstraction was adding a bit too much complexity (e.g. the addition of the `fileData` interface, the `newZipFileData` and `newTarFileData` etc.) for a (not too) complicated function. We could probably do a bit more refactoring, but I left that for later if we want to.

Reply:
The main reason for the abstraction was in fact removing the code duplication, when I was doing it I was envisioning having multiples of different file types, but as we moved along, it's clear that it's very unlikely that we will get support in for more than tar & zip, thus I am good with removing the abstracted layer. 

As you said, if we end up needing more and more file types, then probably would be the time to re-factor and create that additional layer.

Reply:
Yes, it's a pity it wasn't easy to reuse the code, due to reading, and iterating over the files in Tar and Zip archives is quite different